### ISSUE TITLE
Metro is not detecting newly created folders in development builds

### ISSUE THREAD

**Issue by ashuvssut** [expo/expo#21665]
### Summary

**Metro is not detecting newly created folders. I need to restart the bundler after creating a new folder in order to add `.ts/.js/.tsx/.jsx` files in the folder.**

-----------------------
I have a yarn workspace monorepo in which I have my bare-workflow expo app (I followed the expo docs for creating this monorepo)

**I am using development builds**

----------------------
**Explanation: How I found out that **metro is not detecting new folder?****

Whenever I create a new `.ts/.js/.tsx/.jsx` file in `<expo-proj-root>/src/` the app correctly loads the newly created `.ts/.tsx` file *(I can see the "Refreshing..." toast on my app)* 
> Note that `src` folder was already there before starting metro server. That's why there was no error

Now the metro server is running. 
**BUT** whenever I created a new folder (let's say `some-folder`) like `<expo-proj-root>/src/some-folder` and add a file inside it (say `some-folder/abc.ts`) ,  that newly created file is not loading in my app *(there is no "Refreshing..." toast on my app)*
I get this following error:
`Unable to resolve "./src/some-folder/abc" from "App.tsx"`

But when restart the Metro bundler and reload the app, this error is gone.

### Managed or bare workflow?

bare

### What platform(s) does this occur on?

Android

### Package versions

"expo": "~48.0.6",
"expo-dev-client": "~2.1.5",
"react": "18.2.0",
"react-native": "0.71.3",
"typescript": "^4.9.4"

### Environment
```
  expo-env-info 1.0.5 environment info:
    System:
      OS: Linux 5.19 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)
      Shell: 5.8.1 - /usr/bin/zsh
    Binaries:
      Node: 16.19.1 - /run/user/1000/fnm_multishells/6773_1678468726046/bin/node
      Yarn: 1.22.19 - /run/user/1000/fnm_multishells/6773_1678468726046/bin/yarn
      npm: 8.19.3 - /run/user/1000/fnm_multishells/6773_1678468726046/bin/npm
    IDEs:
      Android Studio: AI-212.5712.43.2112.8512546
    npmGlobalPackages:
      expo-cli: 6.3.2
    Expo Workflow: bare
```

### Reproducible demo

I have pushed the minimal repro demo at https://github.com/ashuvssut/expo-dev-client-bug-repro

Read the README given at that repo to get the reproducible steps

### Stacktrace (if a crash is involved)

```
Android Bundling complete 55ms
Unable to resolve "./src/some-folder/abc" from "App.tsx"
 ERROR  [Error: undefined Unable to resolve module ./src/some-folder/abc from /home/ashu/Desktop/walkingpal/apps/client/App.tsx: 

None of these files exist:
  * src/some-folder/abc(.native|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx|.android.js|.native.js|.js|.android.jsx|.native.jsx|.jsx|.android.json|.native.json|.json)
  * src/some-folder/abc/index(.native|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx|.android.js|.native.js|.js|.android.jsx|.native.jsx|.jsx|.android.json|.native.json|.json)
> 1 | import React from "react";
  2 | import { StyleSheet, Text, View } from "react-native";
  3 | import { Button } from "react-native-paper";
  4 | import "expo-dev-client";]
Error: Unable to resolve module ./src/some-folder/abc from /home/ashu/Desktop/walkingpal/apps/client/App.tsx: 

None of these files exist:
  * src/some-folder/abc(.native|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx|.android.js|.native.js|.js|.android.jsx|.native.jsx|.jsx|.android.json|.native.json|.json)
  * src/some-folder/abc/index(.native|.android.ts|.native.ts|.ts|.android.tsx|.native.tsx|.tsx|.android.js|.native.js|.js|.android.jsx|.native.jsx|.jsx|.android.json|.native.json|.json)
> 1 | import React from "react";
  2 | import { StyleSheet, Text, View } from "react-native";
  3 | import { Button } from "react-native-paper";
  4 | import "expo-dev-client";
    at ModuleResolver.resolveDependency (/home/ashu/Desktop/walkingpal/node_modules/metro/src/node-haste/DependencyGraph/ModuleResolution.js:137:15)
    at DependencyGraph.resolveDependency (/home/ashu/Desktop/walkingpal/node_modules/metro/src/node-haste/DependencyGraph.js:260:43)
    at Object.resolve (/home/ashu/Desktop/walkingpal/node_modules/metro/src/lib/transformHelpers.js:177:21)
    at Graph._resolveDependencies (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/Graph.js:432:35)
    at Graph._processModule (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/Graph.js:218:38)
    at async Graph._traverseDependenciesForSingleFile (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/Graph.js:207:5)
    at async Graph.traverseDependencies (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/Graph.js:147:9)
    at async DeltaCalculator._getChangedDependencies (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:247:42)
    at async DeltaCalculator.getDelta (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler/DeltaCalculator.js:90:16)
    at async DeltaBundler.getDelta (/home/ashu/Desktop/walkingpal/node_modules/metro/src/DeltaBundler.js:67:12)
```

**Comment by ashuvssut** [comment]
I can confirm that this issue exists in managed workflow development builds too.

My other vanilla React Native projects(non-expo) work completely fine. So I can confirm that I have my development environment properly setup. I think this issue only lies with expo developement builds (both bare and managed)

**Comment by ashuvssut** [comment]
I have created an relevant duplicate of this issue in react-native repository also. https://github.com/facebook/react-native/issues/36511
